subiquity/server/mounter.py

```diff
@@ -0,0 +1,186 @@
+            fp.write(content)
+
+
+@attr.s(auto_attribs=True, kw_only=True)
+class Mountpoint(_MountBase):
+    mountpoint: str
+
+
+@attr.s(auto_attribs=True, kw_only=True)
+class OverlayMountpoint(_MountBase):
+    # The first element in lowers will be the bottom layer and the last element
+    # will be the top layer.
+    lowers: List["Lower"]
+    upperdir: Optional[str]
+    mountpoint: str
+
+
+Lower = Union[Mountpoint, str, OverlayMountpoint]
+
+
+@functools.singledispatch
+def lowerdir_for(x):
+    """Return value suitable for passing to the lowerdir= overlayfs option."""
+    raise NotImplementedError(x)
+
+
+@lowerdir_for.register(str)
+def _lowerdir_for_str(path):
+    return path
+
+
+@lowerdir_for.register(Mountpoint)
+def _lowerdir_for_mnt(mnt):
+    return mnt.mountpoint
+
+
+@lowerdir_for.register(OverlayMountpoint)
+def _lowerdir_for_ovmnt(ovmnt):
+    # One cannot indefinitely stack overlayfses so construct an
+    # explicit list of the layers of the overlayfs.
+    return lowerdir_for([ovmnt.lowers, ovmnt.upperdir])
+
+
+@lowerdir_for.register(list)
+def _lowerdir_for_lst(lst):
+    return ':'.join(reversed([lowerdir_for(item) for item in lst]))
+
+
+class Mounter:
+
+    def __init__(self, app):
+        self.app = app
+        self.tmpfiles = TmpFileSet()
+        self._mounts: List[Mountpoint] = []
+
+    async def mount(self, device, mountpoint, options=None, type=None):
+        opts = []
+        if options is not None:
+            opts.extend(['-o', options])
+        if type is not None:
+            opts.extend(['-t', type])
+        await self.app.command_runner.run(
+            ['mount'] + opts + [device, mountpoint], private_mounts=False)
+        m = Mountpoint(mountpoint=mountpoint)
+        self._mounts.append(m)
+        return m
+
+    async def unmount(self, mountpoint: Mountpoint, remove=True):
+        if remove:
+            self._mounts.remove(mountpoint)
+        await self.app.command_runner.run(
+                ['umount', mountpoint.mountpoint],
+                private_mounts=False)
+
+    async def setup_overlay(self, lowers: List[Lower]) -> OverlayMountpoint:
+        tdir = self.tmpfiles.tdir()
+        target = f'{tdir}/mount'
+        lowerdir = lowerdir_for(lowers)
+        upperdir = f'{tdir}/upper'
+        workdir = f'{tdir}/work'
+        for d in target, workdir, upperdir:
+            os.mkdir(d)
+
+        options = f'lowerdir={lowerdir},upperdir={upperdir},workdir={workdir}'
+
+        mount = await self.mount(
+            'overlay', target, options=options, type='overlay')
+
+        return OverlayMountpoint(
+            lowers=lowers,
+            mountpoint=mount.p(),
+            upperdir=upperdir)
+
+    async def cleanup(self):
+        for m in reversed(self._mounts):
+            await self.unmount(m, remove=False)
+        self.tmpfiles.cleanup()
+
+
+class DryRunMounter(Mounter):
+
+    async def setup_overlay(self, lowers: List[Lower]) -> OverlayMountpoint:
```

Comment:
We'll probably want to rework this function to make it less coupled with the APT stuff.
Would you add a TODO ?